### ISSUE TITLE
Leave-one-out cross-validation

### ISSUE THREAD

**Issue by puntopasta** [automl/auto-sklearn#149]
So I was considering to implement a leave-one-out option for the nested cross-validation, or more generally leave-X-out. 

The idea is that the user can specify for each data sample from which subject/source it came. Upon fold creation, this information is used to make sure that samples from any source stay confined in one of the folds, making sure that when doing CV the model is not trained on data from the subjects in the fold set.

Are there any intentions to add this feature in the near future that I should wait for, or can I go ahead and make my own implementation?

**Comment by mfeurer** [comment]
While this is an important feature, I'm afraid that we won't have time in the near future to implement this ourselves. In my opinion, one would like to pass an object from [sklearn.cross_validation](http://scikit-learn.org/0.17/modules/classes.html#module-sklearn.cross_validation) to the `AutoSklearnClassifier`. Your usecase would call for the use of [LeavePGroupsOut](http://scikit-learn.org/0.17/modules/generated/sklearn.cross_validation.LeavePLabelOut.html#sklearn.cross_validation.LeavePLabelOut). What do you think about this solution? If you want to implement this, I would be very happy to guide you implementing this.

**Comment by puntopasta** [comment]
The solution that you propose will probably be the most elegant, yes. I will have a look at it this or next month and keep you updated.

**Comment by puntopasta** [comment]
I see that AutoSKLearnClassifier() receives a resampling strategy, which is a string currently. I guess we would create another type which would be something called 'lpgo' or similar? Then as resampling strategy parameters we could introduce two parameters:

- labels: list of labels for each instance denoting it's group
- p: number of groups to leave out

After parsing this input, auto-ml can create an LPGOEvaluator (derives from AbstractEvaluator) which internally will create an sklearn.model_selection.LeavePGroupsOut instance and then implement the abstract functions accordingly.

What do you think?

**Comment by mfeurer** [comment]
I think this is a good solution, but it will make the code more messy. I would rather like if the user can put a CV object into auto-sklearn. Then, there would only be a single evaluator which iterates the states of the CV object. What do you think about this alternative solution?

**Comment by puntopasta** [comment]
Yes, that would work as well, but in the current version that I have checked out there are no objects being passed to AutoSKLearnClassifier() right? I see in the input validation that it expects one of the strings that correspond to some evaluator. You suggest to implement this one differently?

**Comment by mfeurer** [comment]
This is correct, the current version accepts a string and uses an if/else statement from thereon. I suggest to change this part of the code, and rather accept a CV object from scikit-learn.

**Comment by puntopasta** [comment]
Fair enough! So let's say we use sklearn.model_selection.BaseCrossValidator and then use the abstract split() function to generate train/test splits accordingly.

**Step 1:** So I added in automl.AutoML._fit() an extra check on input:  isinstance(self._resampling_strategy, BaseCrossValidator)

**Step 2:** in __init__.eval_with_limits() I added the following:
elif isinstance(resampling_strategy, BaseCrossValidator):
        eval_function = eval_scikitlearn_crossvalidator

**Step 3:** in cv_evaluator.py I now need to implement a function eval_scikitlearn_crossvalidator. Can you give me hints on how to implement this function? I see the other ones take a ton of input variables, are those all needed?

**Comment by mfeurer** [comment]
Great to hear that you're already working on this. First of all, please make sure that your're working against the development branch. This basically requires you to redo **Step 2** and add the check into a class called `ExecuteTaFuncWithQueue`. For the time being, you can simply subclass the `CVEvaluator` and provide a different `get_train_test_split` functionality.

**Comment by cathalgarvey** [comment]
I'd love to see a feature that would allow me to use an arbitrary CV function from `sklearn`. In my case, I'd like to use GroupKFold. Is there any progress on this feature since November? :)

**Comment by puntopasta** [comment]
Sorry, I didn't continue with this in the end. After trying the currently implemented CV on our dataset we didn't see enough reason to invest more time in hyper-parameter optimization.

**Comment by mfeurer** [comment]
@cathalgarvey implementing this is not on my stack for the next few months. However, given the latest restructuring of Auto-sklearn it should be straight-forward to implement. If you want to contribute to Auto-sklearn I can give you a brief tour of the necessary changes in order to be able to submit a PR with this feature.

**Comment by mfeurer** [comment]
This has been implemented via #442. Please reopen if there is any open issue left with respect to custom CV splitters.
